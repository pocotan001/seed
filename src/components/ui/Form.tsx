import { FORM_ERROR, FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

export type IFormOnSubmit = (
  values: IFormValues,
  form: FormApi
) => Promise<void | string> | void | string;

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: IFormOnSubmit;
      }
    > {}

const focusOnErrors = createDecorator();

export default class Form extends React.PureComponent<IFormProps> {
  handleSubmit = async (values: IFormValues, form: FormApi): Promise<any> => {
    const { onSubmit } = this.props;
    const err = await onSubmit(values, form);

    if (err) {
      return { [FORM_ERROR]: err };
    }
  };

  render() {
    const { onSubmit, ...rest } = this.props;

    return (
      <FinalForm
        decorators={[focusOnErrors]}
        onSubmit={this.handleSubmit}
        {...rest}
      />
    );
  }
}
