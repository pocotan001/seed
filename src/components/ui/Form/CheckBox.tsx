import * as React from "react";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import { ITheme } from "~/components/styles/theme";
import styled from "~/components/styles/themedStyledComponents";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";

type IInputAttributes = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "type"
>;

interface IAdditionalProps extends IMarginProps {
  children?: React.ReactNode;
  theme?: ITheme;
}

const CheckBox: React.SFC<
  IFieldRenderProps<IInputAttributes> & IAdditionalProps
> = ({ input, meta, className, children, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return (
    <label className={className}>
      <input {...input} {...rest} aria-invalid={isInvalid} />
      {children && <span className="label">{children}</span>}
    </label>
  );
};

const StyledCheckBox = styled(CheckBox)`
  display: inline-block;
  vertical-align: middle;
  ${margin};

  > input {
    vertical-align: middle;
    margin-right: 8px;
  }

  > .label {
    vertical-align: middle;
  }
`;

const AdaptedCheckBox: React.SFC<
  IFieldProps<IInputAttributes> & IAdditionalProps
> = props => <Field {...props} type="checkbox" component={StyledCheckBox} />;

export default AdaptedCheckBox;
