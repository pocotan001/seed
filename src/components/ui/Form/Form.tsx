import { FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

interface ISubmitErrors {
  [name: string]: string;
}

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: (
          values: IFormValues,
          form: FormApi,
          callback?: (errors?: ISubmitErrors) => void
        ) => Promise<ISubmitErrors | void> | ISubmitErrors | void;
      }
    > {}

const focusOnErrors = createDecorator();
const Form: React.SFC<IFormProps> = props => (
  <FinalForm decorators={[focusOnErrors]} {...props} />
);

export default Form;
