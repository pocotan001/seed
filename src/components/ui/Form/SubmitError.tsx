import { FORM_ERROR } from "final-form";
import * as React from "react";
import { Field, FieldRenderProps } from "react-final-form";
import { IMarginProps } from "~/components/styles/extends/margin";
import Paragraph from "../Paragraph";

type ISubmitErrorProps = IMarginProps;

const render = ({
  meta: { submitError },
  m,
  mt,
  mr,
  mb,
  ml
}: FieldRenderProps & IMarginProps) =>
  submitError ? (
    <Paragraph
      fz={14}
      c="pink500"
      m={m}
      mt={mt}
      mr={mr}
      mb={mb}
      ml={ml}
      role="alert"
    >
      {submitError}
    </Paragraph>
  ) : null;

/**
 * The whole-form submission error returned by `onSubmit` under the `FORM_ERROR` key
 *
 * @example
 * const handleSubmit = () => ({ [FORM_ERROR]: "Failed" });
 *
 * <Form onSubmit={handleSubmit}>
 *   <SubmitError />
 * </Form>
 */
const SubmitError: React.SFC<ISubmitErrorProps> = props => (
  <Field
    {...props}
    name={FORM_ERROR}
    subscription={{ submitError: true }}
    render={render}
  />
);

export default SubmitError;
