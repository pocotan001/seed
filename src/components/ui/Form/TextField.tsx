import * as React from "react";
import styled, { css } from "styled-components";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import { colors } from "~/components/styles/theme";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";

type IInputAttributes = React.InputHTMLAttributes<HTMLInputElement>;
export type ITextFieldStyleProps = IMarginProps;
type IAdditionalProps = ITextFieldStyleProps;

const TextField: React.SFC<
  IFieldRenderProps<IInputAttributes> & IAdditionalProps
> = ({ input, meta, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return <input {...input} {...rest} aria-invalid={isInvalid} />;
};

export const textFieldStyles = css<ITextFieldStyleProps>`
  font-size: 1rem;
  line-height: 1.5;
  display: block;
  padding: 0.4em 0.75em;
  width: 100%;
  max-width: 100%;
  color: ${colors.grey800};
  border: 1px solid ${colors.grey400};
  border-radius: 3px;
  background: ${colors.white};

  &:hover {
    border-color: ${colors.grey500};
  }

  &:disabled {
    cursor: default;
    opacity: 0.6;
    border-color: ${colors.grey400};
  }

  &[aria-invalid="true"] {
    border-color: ${colors.pink500};
  }

  &::-webkit-input-placeholder {
    color: ${colors.grey400};
  }

  &::-moz-placeholder {
    color: ${colors.grey400};
  }

  ${margin};
`;

const StyledTextField = styled(TextField)`
  ${textFieldStyles};
`;

const AdaptedTextField: React.SFC<
  IFieldProps<IInputAttributes> & IAdditionalProps
> = props => <Field {...props} component={StyledTextField} />;

AdaptedTextField.defaultProps = {
  type: "text"
};

export default AdaptedTextField;
