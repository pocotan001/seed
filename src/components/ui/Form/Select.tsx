import * as React from "react";
import { ITheme } from "~/components/styles/theme";
import styled from "~/components/styles/themedStyledComponents";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";
import { ITextFieldStyleProps, textFieldStyles } from "./TextField";

type ISelectAttributes = React.SelectHTMLAttributes<HTMLSelectElement>;

interface IAdditionalProps extends ITextFieldStyleProps {
  theme?: ITheme;
}

const Select: React.SFC<
  IFieldRenderProps<ISelectAttributes> & IAdditionalProps
> = ({ input, meta, children, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  if (rest.multiple) {
    input.value = input.value || [];
    rest.size = Array.isArray(children) ? children.length : 1;
  }

  return (
    <select {...input} {...rest} aria-invalid={isInvalid}>
      {children}
    </select>
  );
};

const StyledSelect = styled(Select)`
  ${textFieldStyles};

  &:not([multiple]):not([size]) {
    /* (font-size * line-height + padding) + border-width */
    height: calc(2.3em + 2px);
  }
`;

const AdaptedSelect: React.SFC<
  IFieldProps<ISelectAttributes> & IAdditionalProps
> = props => <Field {...props} type="select" component={StyledSelect} />;

export default AdaptedSelect;
