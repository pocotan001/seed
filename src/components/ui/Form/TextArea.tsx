import * as React from "react";
import styled from "styled-components";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";
import { ITextFieldStyleProps, textFieldStyles } from "./TextField";

type ITextAreaAttributes = React.TextareaHTMLAttributes<HTMLTextAreaElement>;
type IAdditionalProps = ITextFieldStyleProps;

const TextArea: React.SFC<
  IFieldRenderProps<ITextAreaAttributes> & IAdditionalProps
> = ({ input, meta, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return <textarea {...input} {...rest} aria-invalid={isInvalid} />;
};

const StyledTextArea = styled(TextArea)`
  resize: vertical;
  ${textFieldStyles};
`;

const AdaptedTextArea: React.SFC<
  IFieldProps<ITextAreaAttributes> & IAdditionalProps
> = props => <Field {...props} component={StyledTextArea} />;

export default AdaptedTextArea;
