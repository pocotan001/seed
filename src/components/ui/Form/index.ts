import AutoSave from "./AutoSave";
import CheckBox from "./CheckBox";
import _Form from "./Form";
import Radio from "./Radio";
import Select from "./Select";
import SubmitError from "./SubmitError";
import TextArea from "./TextArea";
import TextField from "./TextField";
import ValidationError from "./ValidationError";

const Form: typeof _Form & {
  AutoSave: typeof AutoSave;
  CheckBox: typeof CheckBox;
  Radio: typeof Radio;
  Select: typeof Select;
  SubmitError: typeof SubmitError;
  TextArea: typeof TextArea;
  TextField: typeof TextField;
  ValidationError: typeof ValidationError;
} = _Form as any;

Form.AutoSave = AutoSave;
Form.CheckBox = CheckBox;
Form.Radio = Radio;
Form.Select = Select;
Form.SubmitError = SubmitError;
Form.TextArea = TextArea;
Form.TextField = TextField;
Form.ValidationError = ValidationError;

export default Form;
