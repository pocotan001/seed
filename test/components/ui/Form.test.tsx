import * as React from "react";
import * as renderer from "react-test-renderer";
import Form from "~/components/ui/Form";

const {
  TextField,
  TextArea,
  Select,
  CheckBox,
  Radio,
  AutoSave,
  ValidationError,
  SubmitError
} = Form;

const handleSubmit = () => undefined;
const save = () => undefined;

describe("<Form>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Form onSubmit={handleSubmit}>{() => "Alo"}</Form>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<TextField>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <TextField name="textfield" />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<TextArea>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <TextArea name="textarea" />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<Select>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => (
            <Select name="select">
              <option value="a">A</option>
              <option value="b">B</option>
              <option value="c">C</option>
            </Select>
          )}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<CheckBox>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => (
            <CheckBox name="checkbox" value="a">
              A
            </CheckBox>
          )}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<Radio>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => (
            <Radio name="radio" value="a">
              A
            </Radio>
          )}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<AutoSave>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <AutoSave onRequestSave={save} />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<ValidationError>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <ValidationError name="a" />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("<SubmitError>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Form onSubmit={handleSubmit}>{() => <SubmitError />}</Form>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
