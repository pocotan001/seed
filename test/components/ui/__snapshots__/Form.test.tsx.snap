// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoSave> renders correctly 1`] = `null`;

exports[`<CheckBox> renders correctly 1`] = `
<label
  className="sc-htpNat cxybjL"
>
  <input
    aria-invalid={false}
    checked={false}
    className="sc-bdVaJa hPFzLn"
    name="checkbox"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="checkbox"
    value="a"
  />
  <span
    className="sc-bwzfXH kpWatx"
  >
    A
  </span>
</label>
`;

exports[`<Form> renders correctly 1`] = `"Alo"`;

exports[`<Radio> renders correctly 1`] = `
<label
  className="sc-EHOje fvYDam"
>
  <input
    aria-invalid={false}
    checked={false}
    className="sc-bxivhb kcGHbr"
    name="radio"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="a"
  />
  <span
    className="sc-ifAKCX fzxfPR"
  >
    A
  </span>
</label>
`;

exports[`<Select> renders correctly 1`] = `
<select
  aria-invalid={false}
  className="sc-gzVnrw hCNSdy"
  name="select"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  value=""
>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
  <option
    value="c"
  >
    C
  </option>
</select>
`;

exports[`<SubmitError> renders correctly 1`] = `null`;

exports[`<TextArea> renders correctly 1`] = `
<textarea
  aria-invalid={false}
  className="sc-iwsKbI hrhgbR"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  value=""
/>
`;

exports[`<TextField> renders correctly 1`] = `
<input
  aria-invalid={false}
  className="sc-bZQynM cdtlqu"
  name="textfield"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  type="text"
  value=""
/>
`;

exports[`<ValidationError> renders correctly 1`] = `null`;
